src/utility/copy.cpp

```diff
@@ -3,30 +3,134 @@
+	core_yyscan_t yyscanner;
+	core_yy_extra_type yyextra;
+	core_YYSTYPE yylval;
+	YYLTYPE yylloc;
+	int token_offset = -1;
+	int nested_level = 0;
+	int tok;
+
+	yyscanner = scanner_init(query, &yyextra, &ScanKeywords, ScanKeywordTokens);
+
+	for (;;) {
+		tok = core_yylex(&yylval, &yylloc, yyscanner);
+		if (tok == 0) {
+			break;
+		} else if (tok == '(') {
+			nested_level++;
+		} else if (tok == ')') {
+			nested_level--;
+		} else if (tok == token_to_match && nested_level == 0) {
+			token_offset = yylloc;
+			break;
+		}
+	}
+	scanner_finish(yyscanner);
+	return token_offset;
+}
+
+static std::string
+CreateRelationCopyString(ParseState *pstate, CopyStmt *copy_stmt, bool *allowed) {
+	ParseNamespaceItem *nsitem;
+#if PG_VERSION_NUM >= 160000
+	RTEPermissionInfo *perminfo;
+#else
+	RangeTblEntry *rte;
+#endif
+	Relation rel;
+	Oid relid;
+	duckdb::string relation_copy;
+
+	/* Open and lock the relation, using the appropriate lock type. */
+	rel = table_openrv(copy_stmt->relation, AccessShareLock);
+	relid = RelationGetRelid(rel);
+	nsitem = addRangeTableEntryForRelation(pstate, rel, AccessShareLock, NULL, false, false);
+
+#if PG_VERSION_NUM >= 160000
+	perminfo = nsitem->p_perminfo;
+	perminfo->requiredPerms = ACL_SELECT;
+#else
+	rte = nsitem->p_rte;
+	rte->requiredPerms = ACL_SELECT;
+#endif
+
+#if PG_VERSION_NUM >= 160000
+	if (!ExecCheckPermissions(pstate->p_rtable, list_make1(perminfo), false)) {
+		ereport(WARNING,
+		        (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+		         errmsg("(PGDuckDB/CreateRelationCopyString) Failed Permission \"%s\"", RelationGetRelationName(rel))));
+		*allowed = false;
+	}
+#else
+	if (!ExecCheckRTPerms(pstate->p_rtable, true)) {
+		ereport(WARNING,
+		        (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+		         errmsg("(PGDuckDB/CreateRelationCopyString) Failed Permission \"%s\"", RelationGetRelationName(rel))));
+		*allowed = false;
+	}
+#endif
+
+	table_close(rel, AccessShareLock);
+
+	if(!*allowed) {
+		return relation_copy;
+	}
+
+	/*
+	 * RLS for relation. We should probably bail out at this point.
+	 */
+	if (check_enable_rls(relid, InvalidOid, false) == RLS_ENABLED) {
+		ereport(WARNING,
+		        (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+		         errmsg("(PGDuckDB/CreateRelationCopyString) RLS enabled on \"%s\"", RelationGetRelationName(rel))));
+		*allowed = false;
+		return relation_copy;
+	}
+
+	relation_copy += pgduckdb_relation_name(relid);
+	if (copy_stmt->attlist) {
+		ListCell *lc;
+		relation_copy += "(";
+		bool first = true;
+		foreach (lc, copy_stmt->attlist) {
+			if (!first) {
+				relation_copy += ", ";
+			}
+			first = false;
+			relation_copy += strVal(lfirst(lc));
+		}
+		relation_copy += ") ";
+	}
+
+	*allowed = true;
+	return relation_copy;
+}
+
 bool
 DuckdbCopy(PlannedStmt *pstmt, const char *query_string, struct QueryEnvironment *query_env, uint64 *processed) {
 	CopyStmt *copy_stmt = (CopyStmt *)pstmt->utilityStmt;
```

Comment:
Not really related to this PR. But we should probably check `copy_stmt->filename` for NULL here.